tsfc/kernel_interface/themis.py

```diff
@@ -0,0 +1,411 @@
+        elif integral_type in ['interior_facet', 'interior_facet_vert']:
+            facet = gem.Variable('facet', (2,))
+            self._entity_number = {
+                '+': gem.VariableIndex(gem.Indexed(facet, (0,))),
+                '-': gem.VariableIndex(gem.Indexed(facet, (1,)))
+            }
+        # KEY: swapped compared to Firedrake
+        elif integral_type == 'interior_facet_horiz':
+            self._entity_number = {'+': 0, '-': 1}
+
+    def set_arguments(self, arguments, multiindices):
+        """Process arguments.
+
+        :arg arguments: :class:`ufl.Argument`s
+        :arg multiindices: GEM argument multiindices
+        :returns: GEM expression representing the return variable
+        """
+        self.local_tensor, expressions = prepare_arguments(
+            arguments, multiindices, self.scalar_type, interior_facet=self.interior_facet)
+        return expressions
+
+    def set_coordinates(self, domain):
+        """Prepare the coordinate field.
+
+        :arg domain: :class:`ufl.Domain`
+        """
+        # Create a fake coordinate coefficient for a domain.
+        f = Coefficient(FunctionSpace(domain, domain.ufl_coordinate_element()))
+        self.domain_coordinate[domain] = f
+        self.coordinates_arg = self._coefficient(f, "coords")
+
+    def set_coefficients(self, integral_data, form_data):
+        """Prepare the coefficients of the form.
+
+        :arg integral_data: UFL integral data
+        :arg form_data: UFL form data
+        """
+        coefficients = []
+        coefficient_numbers = []
+        # enabled_coefficients is a boolean array that indicates which
+        # of reduced_coefficients the integral requires.
+        for i in range(len(integral_data.enabled_coefficients)):
+            if integral_data.enabled_coefficients[i]:
+                coefficient = form_data.function_replace_map[form_data.reduced_coefficients[i]]
+                if type(coefficient.ufl_element()) == ufl_MixedElement:
+                    split = [Coefficient(FunctionSpace(coefficient.ufl_domain(), element))
+                             for element in coefficient.ufl_element().sub_elements()]
+                    coefficients.extend(split)
+                    self.coefficient_split[coefficient] = split
+                else:
+                    coefficients.append(coefficient)
+                # This is which coefficient in the original form the
+                # current coefficient is.
+                # Consider f*v*dx + g*v*ds, the full form contains two
+                # coefficients, but each integral only requires one.
+                coefficient_numbers.append(form_data.original_coefficient_positions[i])
+        for i, coefficient in enumerate(coefficients):
+            self.coefficient_args.append(
+                self._coefficient(coefficient, "w_%d" % i))
+        self.kernel.coefficient_numbers = tuple(coefficient_numbers)
```

Comment:
Again, lots of this code is identical to the firedrake kernel builder.

Reply:
There is no longer a Themis interface, instead the Firedrake interface (with tabulations added) works for us.